### ISSUE TITLE
Depagenation

### ISSUE THREAD

**Issue by oxinabox** [JuliaCloud/AWS.jl#340]
all AWS results are pagenated.
You need to access the `NextToken"` value from the result and rerun the query to get the next one.
Which really sucks to use.

I have a simple function that one can use to depagenate things
```julia
function depagenate(f, key, args...; params...)
    d_params = Dict{String, Any}(string(k)=>v for (k,v) in pairs(params))
    resp = f(args..., d_params)
    gathered = resp[key]
    while haskey(resp, "nextToken")
        d_params["nextToken"] = resp["nextToken"]
        resp = f(args..., d_params)
        append!(gathered, resp[key])
    end
    return gathered
end
```

usage example
```julia
jobs = depagenate(
    Sagemaker.list_training_jobs_for_hyper_parameter_tuning_job,  # the function
    "TrainingJobSummaries",  # the key that has the data to be concatenated
    "MyHyperTuningJob-1",    # the function's arguments
)
```

Its not super smart. A better one would use `async` and `Channels` to run fetching in the backround, so that you can start consuming them before they all arrive.

That assumes the token is called `nextToken` which is not always true.
@mattBrzezinski  says
 
> S3 for instance uses nextContinuationToken
> - https://github.com/aws/aws-sdk-js/blob/master/apis/s3-2006-03-01.normal.json#L5998
> So we would need to:
> - On a service-by-service level find which endpoints outputs have a NextToken shape (these are pagination endpoints)
> - Create functionality to paginate requests based off of the name of this token

So what we want to do is automatically generate the pagenated version of each function, which uses that key.
and has a smart way of combing the results.
Probably appending vectors, and vcatting nonvectors, and dropping the pagenation token.

**Comment by mattBrzezinski** [comment]
Dealing with differently named continuation tokens isn't too much of a problem. When we are generating services we can generate another mapping:

```julia
# services/sagermaker.jl
const DEPAGINATION = Dict{
    list_training_jobs_for_hyper_parameter_tuning_job => "nextToken"
}
```

One tricky part here as per usual is AWS' inconsistency with naming. Not only is the name of the response item `nextToken, nextContinuationToken` inconsistent, so is the shape name. For instance S3 and SageMaker call it [NextToken](https://github.com/aws/aws-sdk-js/blob/0ea1cbd1cd04e2c5840dba36fe4a44c734448057/apis/s3-2006-03-01.normal.json#L6028), while CognitoIdentity calls its shape [PaginationKey](https://github.com/aws/aws-sdk-js/blob/0ea1cbd1cd04e2c5840dba36fe4a44c734448057/apis/cognito-identity-2014-06-30.normal.json#L1412).

This largely isn't that big of a deal, but it means that we cannot really full automate this as some services will be missed if they differ. What we can do here is, note very pseudo code,

```julia
function generate_high_level_functions(data)
    PAGINATION_KEY_NAMES = ["NextToken", "PaginationKey"]
    paginated_functions = []

    for keys in data[output_shape]
        if key in PAGINATION_KEY_NAMES
            push!(paginated_functions, keys)
        end
    end
end
```

**Comment by iamed2** [comment]
The paginators should be listed in the paginators.json file, with the keys needed to parse them. I don't think the shape _name_ actually matters for the JSON APIs. 

I think it would make sense to have an iterator:

```
struct Paginator
    requester::Function  # closure accepting just the relevant pagination args
    input_token::String
    output_token::String
    result_key::String
    limit_key::String
    limit::Int
    responses::Vector{...}
end
```

Storing the responses in a vector would allow random access, how iterators in Julia usually work.

**Comment by mattBrzezinski** [comment]
> The paginators should be listed in the paginators.json file, with the keys needed to parse them. I don't think the shape _name_ actually matters for the JSON APIs.

Oh right! Thanks for the reminder I forgot that there was a paginators JSON file!

**Comment by oxinabox** [comment]
>  Storing the responses in a vector would allow random access, how iterators in Julia usually work.

For a final user perspective i do not think we should be returning a iterator of responses.
Most (all?) things that are pagenated in AWS are things that each response is a collection of details.
e.g. `Sagemaker.list_training_jobs_for_hyper_parameter_tuning_job` returns a vector of 10  items under `TrainingJobSummaries` in each response.
`Cloudwatch_Logs.filter_log_events` returns a vector of items under then `events` key.

The user wants to end up iterating training job summaries, or events.
Not responses then within the response those items.

**Comment by iamed2** [comment]
Yes, sorry, that is what I meant to imply. You can store each response then parse them when iterating. The `result_key` says which field of the response contains the list of items. You can `mod` to find out where a given ordinal appears in the list of responses.

**Comment by iamed2** [comment]
> I've found that blows through the default exponential retry and gets rate limited pretty hard, and I have to introduce a manual 0.5 second sleep between pages to reliably get by.

What exactly happens with the default retry? Is it running past a maximum number of retries and erroring before you can get your result? Can you see what happens when using https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/logs.html#CloudWatchLogs.Paginator.DescribeLogStreams ?
